### PR TITLE
#160452475 Sign up api verification email 

### PR DESCRIPTION
**Pull request by hnobi**
#### What does this PR do?
Adds signup API controller, middleware and email verification function

#### Description of Task to be completed?
- Get sendGrid token
- Adds signup controller
- Adds signup route
- adds swagger documentation
- Adds signup validation middleware
- Adds test for signup controller and middleware
- Adds  function to generate one time token
- Adds a class for sending emails with sendGrid
#### How should this be manually tested?
- run `npm install` in the terminal  and run `localhost:3000/api/v1/auth/signup` on Postman
- check email inbox for verification link and copy the token e.g `nwmr7d3e7zgnwmr7d3e7zg` and then
-  run `localhost:3000/api/v1/auth/confirmation/nwmr7d3e7zgnwmr7d3e7zg`
#### Any background context you want to provide?
- When testing this feature, use Gmail  as sendGrid can't send mail to Yahoo mail
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/160452475
#### Screenshots (if appropriate)
Nil
#### Questions:
Nil

### CODE REVIEW

File: services/Mail.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @return {json} Returns json object
+   * @memberof Mail
+   */
+  static sendVerification(req, res, next) {
+    const { email, token } = req.user;
+    const url = `http://localhost:3000/api/v1/auth/confirmation/${token}`;
+    const msg = emailMesseges.signupVerification(email, url);
```
Comment (Alameen688) [inline services/Mail.js]: You might want to change  "emailMesseges" to "emailMessages"

File: services/Mail.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+class Mail {
+  /**
+   * @static
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @return {json} Returns json object
+   * @memberof Mail
+   */
+  static sendVerification(req, res, next) {
```
Comment (Alameen688) [inline services/Mail.js:24]: Separation of concerns here is on point. However, this method is not really reusable. We can have it as a "sendMail" function that takes in 2 arguments: the message to be sent and the email it needs to be sent to.
This is just to make it possible for anyone also performing an action related to sending of email to find it easy to reuse such method.

File: services/Mail.js (hunk 1 above)
Comment (hnobi) [reply services/Mail.js]: Thanks, I will change it now

**Review by Alameen688 — APPROVED** [review]
Looks good to me

File: services/Mail.js (hunk 2 above)
Comment (hnobi) [reply services/Mail.js:24]: Thanks for the feedback, I have modularised it now.

Comment (Alameen688) [reply services/Mail.js:24]: Alright, cool. Well-done.

File: services/Mail.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    const { email, token } = req.user;
+    const url = `http://localhost:3000/api/v1/auth/confirmation/${token}`;
```
Comment (blackshady) [inline services/Mail.js]: Nice job on your work so far. I think it is better we have this `url` in an environment variable.
You can name it `BASE_URL = localhost:3000/api/v1/` in your .env file and declear it in this file as `const BASE_URL = process.env.BASE_URL`. 
To use it `const url =`http://BASE_URL/auth/confirmation/${token}` `

**Review by Alameen688 — APPROVED** [review]
Well-done! Feedback has been implemented, looks good to me.

**Pull request merged**